pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp

```diff
@@ -382,159 +260,334 @@ TEST_F(PilzJointTrajectoryControllerTest, testDoubleRequest)
+  EXPECT_NEAR(ros::Time::now().toSec(), start_hold_time.toSec() + STOP_TRAJECTORY_DURATION, TIME_COMPARISON_TOLERANCE_SEC);
 
-  // Now we should have stopped
+  updateController();
+
+  // Second hold should not sleep
+  start_hold_time = ros::Time::now();
   ASSERT_TRUE(controller_->handleHoldRequest(req, resp));
   EXPECT_TRUE(resp.success);
-  controller_->update(ros::Time::now(), ros::Duration(SMALL_PERIOD));
+  EXPECT_NEAR(ros::Time::now().toSec(), start_hold_time.toSec(), TIME_COMPARISON_TOLERANCE_SEC);
+
+  // Now we should have stopped
+  ros::Duration period{DEFAULT_UPDATE_PERIOD_SEC};
+  progressInTime(period);
+  updateController();
   EXPECT_FALSE(controller_->is_executing());
 }
 
 /**
- * @brief Test the is_executing method of the PilzJointTrajectoryController.
+ * @brief Check that a fake start (simply setting state to RUNNING) doesn't trigger that is_executing() returns true.
  *
- * This test is analogue to the test of the is_executing callback.
+ * Increases line coverage.
+ */
+TEST_F(PilzJointTrajectoryControllerTest, testFakeStart)
+{
+  controller_->state_ = controller_->RUNNING;
+  EXPECT_FALSE(controller_->is_executing()) << "Controller is executing unexpectedly";
+}
+
+/**
+ * @brief Test increases function coverage by ensuring that all Dtor variants are called.
+ */
+TEST_F(PilzJointTrajectoryControllerTest, testD0Destructor)
+{
+  ControllerPtr controller {new Controller()};
+}
+
+
+///////////////////////////////////////////////
+//    Parameterized tests for IsExecuting    //
+///////////////////////////////////////////////
+
+typedef std::function<testing::AssertionResult(const ControllerPtr&, bool&)> InvokeIsExecuting;
+
+static testing::AssertionResult InvokeIsExecutingMethod(const ControllerPtr& controller, bool& result)
+{
+  result = controller->is_executing();
+  return testing::AssertionSuccess();
+}
+
+static testing::AssertionResult InvokeIsExecutingServiceCallback(const ControllerPtr& controller, bool &result)
+{
+  std_srvs::TriggerRequest req;
+  std_srvs::TriggerResponse resp;
+  if(!controller->handleIsExecutingRequest(req, resp))
+  {
+    return testing::AssertionFailure() << "Callback of is_executing returned false unexpectedly.";
+  }
+
+  result = resp.success;
+  return testing::AssertionSuccess();
+}
+
+/**
+ * @brief For testing both the isExecuting method and the is_executing service callback we use parameterized tests.
+ */
+class PilzJointTrajectoryControllerIsExecutingTest : public PilzJointTrajectoryControllerTest,
+                                                     public testing::WithParamInterface<InvokeIsExecuting>
+{
+protected:
+  testing::AssertionResult invokeIsExecuting();
+  testing::AssertionResult waitForIsExecutingResult(bool expectation);
+
+  /**
+   * @brief Perform init, start, unhold and update, such that controller is ready for executing.
+   */
+  testing::AssertionResult performFullControllerStartup();
+
+  bool is_executing_result;
```

Comment:
I added a comment as an explanation:

> //! The return value indicates if the call was successful (in case of a service callback), the actual result
//! of the is-executing-function is assigned (via reference) to the second argument.

Reply:
In other words, this is used to distinguish different outcomes of the service callback.